components/base/wheeled/sensorbase.go

```diff
@@ -0,0 +1,327 @@
+	atTarget = math.Abs(targetYaw-currYaw) < errorBound
+	overShot = hasOverShot(currYaw, startYaw, targetYaw, dir)
+	minTravel = math.Abs(currYaw-startYaw) > math.Abs(angleDeg*increment)
+	return atTarget, overShot, minTravel
+}
+
+func getCurrentYaw(ctx context.Context, ms movementsensor.MovementSensor,
+) (float64, error) {
+	ctx, done := context.WithCancel(ctx)
+	defer done()
+	orientation, err := ms.Orientation(ctx, nil)
+	if err != nil {
+		return 0, errors.Wrap(
+			err, "error getting orientation from sensor, spin will proceed without sensor feedback",
+		)
+	}
+	// Add Pi  to make the computation for overshoot simpler
+	// turns imus from -180 -> 180 to a 0 -> 360 range
+	return addAnglesInDomain(rdkutils.RadToDeg(orientation.EulerAngles().Yaw), 0), nil
+}
+
+func addAnglesInDomain(target, current float64) float64 {
+	angle := target + current
+	// reduce the angle
+	angle = math.Mod(angle, oneTurn)
+
+	// force it to be the positive remainder, so that 0 <= angle < 360
+	angle = math.Mod(angle+oneTurn, oneTurn)
+	return angle
+}
+
+func findSpinParams(angleDeg, degsPerSec, currYaw float64) (float64, float64, int) {
+	targetYaw := addAnglesInDomain(angleDeg, currYaw)
+	dir := 1.0
+	if math.Signbit(degsPerSec) != math.Signbit(angleDeg) {
+		// both positive or both negative -> counterclockwise spin call
+		// counterclockwise spin calls add allowable angles
+		// the signs being different --> clockwise spin call
+		// cloxkwise spin calls subtract allowable angles
+		dir = -1
+	}
+	fullTurns := int(math.Abs(angleDeg)) / int(oneTurn)
+
+	return targetYaw, dir, fullTurns
+}
+
+// this function does not wrap around 360 degrees currently.
+func angleBetween(current, bound1, bound2 float64) bool {
+	if bound2 > bound1 {
+		inBewtween := current >= bound1 && current < bound2
+		return inBewtween
+	}
+	inBewteen := current > bound2 && current <= bound1
+	return inBewteen
+}
+
+func hasOverShot(angle, start, target, dir float64) bool {
+	switch {
+	case dir == -1 && start > target:
+		// for cases with a quadrant switch from 1 -> 4
+		// check if the current angle is in the regions before the
+		// target and after the start
+		over := angleBetween(angle, target, 0) || angleBetween(angle, 360, start)
+		return over
+	case dir == -1 && target > start:
+		// the overshoot range is the inside range between the start and target
+		return angleBetween(angle, target, start)
+	case dir == 1 && start > target:
+		// for cases with a quadrant switch from 1 <-> 4
+		// check if the current angle is not in the regions after the
+		// target and before the start
+		over := !angleBetween(angle, 0, target) && !angleBetween(angle, start, 360)
+		return over
+	default:
+		// the overshoot range is the outside range between the start and target
+		return !angleBetween(angle, start, target)
+	}
+}
+
+func (s *sensorBase) MoveStraight(
+	ctx context.Context, distanceMm int, mmPerSec float64, extra map[string]interface{},
+) error {
+	s.stopSensors()
```

Comment:
why do we need to call this in every method? can we `defer s.stopSensor()` in Spin and leave `s.stopSensors` in other calls?

Reply:
I cannot cancel the context for that loop to stop otherwise from another call being sent while it is active.  I can't use the API ctx, which is short-lived and cancels the loop. When I cancel the context, I need to spawn a new context with cancel that the spin loop uses

Reply:
The only other thing I can do here, is have a long-running single context that is only made in the constructor and only closed in the closer, and a value on the struct with a mutex that sets a state of whether we should be listening to the sensor in the for loop, and ignoring it if we've called anything other than `Spin`. 

That kind of follows the way the encoded motor does it, but we might end up with more values to check on the struct when we add more sensors to this thing. We should keep values we check on the sensors struct minimal that way.